### PR TITLE
improvement(directives): add v-same-dom

### PR DESCRIPTION
**Pull request by Vgbire**
Please make sure these boxes are checked before submitting your PR, thank you!

* [x] Make sure you follow Element's contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
* [x] Make sure you are merging your commits to `dev` branch.
* [x] Add some descriptions and refer to relative issues for your PR.

### CODE REVIEW

**Comment by JeremyWuuuuu** [comment]
Also modify `el-message-box`

**Comment by Vgbire** [comment]
> Also modify `el-message-box`

OK,I think maybe I should modify the overlay to optimize it.you do the same thing in it.I will make a new pull requrest after.

**Comment by Vgbire** [comment]
@JeremyWuuuuu OK.I didn't know at first that overlay should be optimized instead of changing other business components.

File: packages/dialog/src/index.vue [hunk 1]
```diff
@@ -11,11 +11,13 @@
         :mask="modal"
         :overlay-class="modalClass"
         :z-index="zIndex"
-        @click="onModalClick"
+        :close="onModalClick"
```
Comment (JeremyWuuuuu) [inline packages/dialog/src/index.vue:14]: we use it as a event emitter, why would you change this as a value?

**Comment by JeremyWuuuuu** [comment]
Seemed to me that this is way more complicated than I expected, I think.
I think the more appropriated version should be a hook like this:

### useClickSameDom.ts
```typescript
export default (emitter: () => void) {
    let mousedownTarget = false
    let mouseupTarget = false
    // refer to this https://javascript.info/mouse-events-basics
    // events fired in the order: mousedown -> mouseup -> click
    // we need to set the mousedown handle to false after click
    // fired.
    const onClick = (e: MouseEvent) => {
      // due to these two value were set only when props.mask is true
      // so there is no need to do any extra judgment here.
      // if and only if
      if (mousedownTarget && mouseupTarget) {
        emitter()
      }
      mousedownTarget = mouseupTarget = false
    }
    return onClick
} 

```

**Comment by Vgbire** [comment]
I'm not sure if this change meets your expectations, if not, please let me know.

**Pull request closed without merging**